### ISSUE TITLE
How can we prevent the __typename field being added to generated queries?

### ISSUE THREAD

**Issue by adrian-baker** [apollographql/apollo-kotlin#1458]

**Comment by sav007** [comment]
There is no way to do this as it very critical field to correctly resolve fragments, inline fragments, normalized cache etc.

**Comment by adrian-baker** [comment]
The code generator knows the typename at generation type though, so technically there is no need to retrieve it every time at runtime, is there?

We have a query with a large list of small entries, for which repeating typename for each has a large impact on the resulting payload size (and serialization/deserialization cost).

**Comment by sav007** [comment]
>The code generator knows the typename at generation type though, so technically there is no need to retrieve it every time at runtime, is there?

no code gen knows about the type name it request fields on, but it can be an interface or union:
https://graphql.org/learn/schema/#interfaces so the concrete typename comes with the response.

**Comment by adrian-baker** [comment]
1. In that case there is no need for `__typename` if the type is not a union or an interface, yet it appears to be added for all types.

2. Even for the case of a interface/union, the return value can safely be deserialized into a higher level type without knowing the concrete type name, so even in this case I don't see that `__typename` is necessary. The query can only issue fields in a subtype if they do a fragment with `... on`, specifying explicitly the type, which means the additional `__typename` field is still redundant.

**Comment by sav007** [comment]
Here is a use case:
```
query TestQuery {
  hero {
    ...HeroDetails
  }
}

fragment HumanDetails on Human {
  name
}
```

Requested field `hero` is type of interface `Character`, so in the response we can expect for example another concrete type of this interface `Droid` and as our fragment defined on `Human` we shouldn't parse it. To proper handle this we need to understand what type is returned by requesting `__typename`. 

I can agree that we shouldn't probably request `__typename` for every fields but rather than on those having fragments and inline fragments, I don't remember why it was decided to have them everywhere maybe just to be consistent but I can try to dig into and see what I can do.

Does this make sense to you?

**Comment by adrian-baker** [comment]
Right, yes, thanks for explaining. Though - a minor point - in the example above, you can tell by the presence of the name field in the response (if name is present, it's a Human). I can see though that there will be other cases where this sort of rule isn't possible.

In any case, not adding `__typename` for non fragments would certainly resolve our issue (we're not using them for the problem query).

**Comment by sav007** [comment]
After working on the PR I realized that removing `__typename` field generation will break normalized cache as it totally depends on this field, as any other query with inline / regular fragment can't be resolved from cached response as `__typename` fields is missing. Closing this issue as won't fix.

**Comment by adrian-baker** [comment]
What if caching is not used? It seems inefficient if apollo client is adding a lot of overhead (ie high volume of additional fields serialized) for features not used (caching, polymorphism).

**Comment by sav007** [comment]
Alright, alright we can make it configurable.

**Comment by SnowyCoder** [comment]
Just throwing my two cents: Even if the name doesn't seem like much it can be quite heavy if the site is downloading an array of data (for example readings from a sensor to build a graph).

**Comment by martinbonnin** [comment]
I'm curious how much network bandwidth is actually saved by removing the __typename fields. My guess is that  these fields would gzip very well so in the case of an array, it will run-length-encode all the identical __typename fields and the overhead would be very small.

**Comment by adrian-baker** [comment]
The real overhead is not bandwidth, but serialization and deserialization cost.

**Comment by martinbonnin** [comment]
Heads up that [Apollo Android 3 is now alpha](https://github.com/apollographql/apollo-android/releases/tag/v3.0.0-alpha01) and doesn't add `__typename` by default. It still adds them when there are fragments (inline or named) but that's something that will be way harder or even impossible to remove if we want to be able to parse polymorphic fields.

**Comment by martinbonnin** [comment]
Keeping this open as we can still remove the `__typename` for fields with fragments on object types. But for a lot of cases, `__typename` should appear way less often now. If you have a small object where `__typename` adds too much overhead, you can remove fragments to workaround this.

**Comment by martinbonnin** [comment]
Latest SNAPSHOTs include a `"addTypename"` option to have more fine-grained control over when `__typename` is needed.

```kotlin
apollo {
  /**
   * Add '__typename' for every selection set that contains fragments (inline or named) (Default)
   * 
   * This is adding a lot more '__typename' than the above 'ifPolymorphic" and will be removed in 
   * a future version. If you require '__typename' explicitely, you should add it to your queries.
   **/
  addTypename.set("ifFragments")

  /**
   * Add '__typename' for polymorphic fields, i.e. fields that contains a subfragment 
   * (inline or named) whose type condition isn't a super type of the field type.
   * If a field is monomorphic, no '__typename' will be added.
   * 
   * Note: It also adds '__typename' on fragment definitions that satisfy the same property because fragments
   * could be read from the cache and we don't have a containing field in that case.
   **/
  addTypename.set("ifPolymorphic")

  /**
   * Add '__typename' for abstract fields, i.e. fields that are of union or interface type
   * 
   * Note: It also adds '__typename' on fragment definitions that satisfy the same property because fragments
   * could be read from the cache and we don't have a containing field in that case.
   **/
  addTypename.set("ifAbstract")
}
```

See https://github.com/apollographql/apollo-kotlin/pull/3939 for more details 

At this point I don't think we can remove any more `__typename` without losing some type safety so unless there are strong objections, I'll close this issue with the next release.

**Comment by martinbonnin** [comment]
Closing this ~3 years old issue hopefully for good this time. All the options above are available in `3.2.0`
